### ISSUE TITLE
Inconsistent typing for headers through the project

### ISSUE THREAD

**Issue by Ousret** [urllib3/urllib3#3047]
### Subject

When experimenting in #3030 I did find myself surprised to receive headers as bytes in putheader() while the typing said otherwise.

The project, in the v2.0.x currently accepts headers as follows:

```python
headers: typing.Mapping[str, str] | None = None,
```

I did find 27 occurrences of **headers** as an argument. Only twice did I find that it allows bytes (in response.py).

```python
headers: typing.Mapping[str, str] | typing.Mapping[bytes, bytes] | None = None,
```

But there is a clear indication that the project is supposed to support headers (input) as:

```python
headers: typing.Mapping[str | bytes, str | bytes] | None = None,
```

See `test_user_agent_non_ascii_user_agent`, `test_user_agent_header_not_sent_twice`,

also in requests

`test_header_keys_are_native`, `test_header_validation`

So, in conclusion, is this something urllib3 is supposed to support? If yes we should update the project typing for headers so that we would be warned of potential miss with mypy.

**Comment by sethmlarson** [comment]
Thanks for reporting this @Ousret, I agree that we should use the headers definition of strings and bytes. I'm adding a bounty of $100 to this issue for the fix you describe.

**Comment by sabdulmajid** [comment]
I agree that this is an issue, and I would be willing to work on it if the issue is still there. Is it still something that needs to be fixed @sethmlarson? I'm not an advanced programmer or anything, but I can dabble around and try to get it fixed if need be. Let me know!

**Comment by pquentin** [comment]
@sabdulmajid sure, go ahead. Would a week be enough? @tushar5526 also expressed interest on our Discord.

**Comment by pquentin** [comment]
Assigning to @tushar5526 given the lack of response. Thanks

**Comment by sabdulmajid** [comment]
Oh no! I apologize for my lack of response! I completely forgot to send my message back. You can assign it to me and I will be able to get it done as soon as I can.

**Comment by sabdulmajid** [comment]
Please let me know if I should work on it...once again I apologize for not responding. I had wrote up a response but forgot  to send it! So sorry about that.

**Comment by sabdulmajid** [comment]
Okay, so if I'm not mistaken, I only really have to go through all of the src/ files and review the lines where it says
```python
headers: typing.Mapping[str, str] | None = None,
```
and just change them to
```python
headers: typing.Mapping[typing.Union[str, bytes], typing.Union[str, bytes]] | None = None
```
Am I correct or am I missing something here? According to the OP, they found that 'headers' occurred 27 times throughout the code. On using the "find" tool in VSC, I've found the line 22 times in the src/ folder. I'm assuming that some of the other 'headers' have different definitions for them, not the default one above?

And then I can test it with something like this:
```python
import urllib3

def test_headers():
  headers = {"foo": "bar", "baz": b"quux"}

  client = urllib3.PoolManager()
  response = client.request("GET", "/", headers=headers)

  assert response.headers["foo"] == "bar"
  assert response.headers["baz"] == b"quux"
```

**Comment by tushar5526** [comment]
I was looking into this a bit more. To add support for both strings and bytes, it seems we would have to update HTTPHeaderDict and all the associated methods for `mypy` to not error out. 

For example, 

```
    def __getitem__(self, key: str) -> str:
        val = self._container[key.lower()]
        return ", ".join(val[1:])
```

**Comment by Tinche** [comment]
I've picked this up, and this is actually proving to be kind of tricky. It feels important for urllib3 to be able byte headers so I persevered, but given the fact urllib3 has made it this far without explicit support suggests it's an advanced use case so it doesn't have to be super easy to use.

For requests, from what I see in CPython's codebase, we can technically support bytes alongside strings in header names and values (although [typeshed](https://github.com/python/typeshed/blob/main/stdlib/http/client.pyi#L179) is wrong, we might want to fix that too).

For responses, after going through the codebase, docs and tests I'm having trouble understanding how exactly urllib3 is supposed to return binary headers.

The main problem though: `typing.Mapping` isn't covariant in its key parameter. That's a fancy way of saying that `Mapping[str, T]` isn't a subtype of `Mapping[str | bytes, T]` even though `str` is a subtype of `str | bytes`. This extends to `HTTPHeaderDict` since it's a subtype of `Mapping[str, str]`.

If I change the signature of `PoolManager(headers=)` or `PoolManager.request(headers=)` to `Mapping[str | bytes]`, they don't accept `HTTPHeaderDict` any more.

If I change `HTTPHeaderDict` to be a `Mapping[str | bytes, str | bytes]`, I think the ergonomics will suffer when handling responses (cuz now users will have to check if their response header value is a string or bytes, when in reality it'll never be bytes).

A bonus problem, this code won't work:

```
headers = {"a": "a"}

PoolManager.request(..., headers=headers)  # Mypy: dict[str, str] is incompatible with Mapping[str | bytes, str | bytes]
```

So I think a solution will need to be less naive than just changing a few type annotations.

How constrained are we by backwards compatibility here? I'm guessing _substantially_.

**Comment by Tinche** [comment]
A small update: I managed to land https://github.com/python/typeshed/pull/10978

**Comment by sethmlarson** [comment]
@Tinche Awesome! Happy to hear you're picking this up. I'll assign you since you've landed a PR elsewhere as a part of this work. Let us know if you decide not to move forward later :)

**Comment by August753** [comment]
Hi maintainers, is this issue still open for external contributors?

I noticed it is still open and labeled with the $100 bounty, but it has been assigned for a while. I have a patch that updates urllib3's request header typing to support `str`/`bytes` header names and values without breaking `dict[str, str]` or `HTTPHeaderDict` usage, with tests and mypy passing.

Would you be open to a PR for this?

**Comment by Tinche** [comment]
Feel free to take this over as far as I'm concerned! Sorry, it completely slipped my mind.

**Comment by Rainash48** [comment]
Picking this up in #5051. The patch aligns the public header input typing with the existing `str`/`bytes` runtime support and adds regression coverage for byte-keyed/byte-valued request headers.
